lcm-cmake/lcmUtilities.cmake

```diff
@@ -329,7 +331,10 @@ function(lcm_wrap_types)
           _lcm_add_go_outputs(_GO_SOURCES ${_package_dir}/${_type}.go)
         endif()
         if(DEFINED _JAVA_SOURCES)
-          _lcm_add_outputs(_JAVA_SOURCES ${_package_dir}/${_type}.java)
+          if(_package_dir STREQUAL "")
+            set(_package_dir "lcmtypes")
+          endif()
+            _lcm_add_outputs(_JAVA_SOURCES ${_package_dir}/${_type}.java)
```

Comment:
BTW, accidental change of indent?

Reply:
The reason for implementing it for Java only is because I wanted the behaviour to be consistent with `lcm-gen`. There's no C# support in `lcmUtilities.cmake` and as for go (lcm 1.4) I get _Not yet implemented: go-default-package_. That leaves only Java.

I can split it into two PR's if you wish. What do you prefer?

Oh, the indentation was accidentally left behind after making some final changes. I'll fix that.

Reply:
Okay, I guess I can buy that. I'd feel better if someone more familiar with LCM can also weigh in, though. (@ashuang maybe?) I mostly know the CMake bits, and that's about it :wink:.

BTW, if the issue is that certain languages require a package name, maybe it would make sense to move this logic up to line 328? And add a comment e.g. "Following languages require a package name". Having it inside a check for a language being enabled might lead to unexpected behavior, if it is intended to apply to multiple languages.

(I'm not familiar with Go, so don't know what's going wrong there...)

Reply:
Oops, I saw this comment now.

The default package for C# is different from java's default package, so there needs to be language checks anyway.
I guess it's cleaner to move them out as you suggest, but with only java being able to use a default package at the moment (unless that's changed since 1.4; can't try it since I'm on my phone), the refactoring would basically mean adding an additional check for java and not much else. Shouldn't be too bad to wait with the refactoring till C# wrapper generation is implemented in `lcmUtilities.lcm` and/or `lcm-gen` uses the default package switch for Go, I suppose.

Reply:
Ack. Well, the problem is, if they're different, then the current logic will have changed the package for anything downstream of Java. (Which is nothing, but that can't be true for all languages.) So I think instead you would need to have a language-specific package name that is used, or else tweak the logic to know to unset the package name again after calling `_lcm_add_outputs`. Possibly something as simple as:
```cmake
if(_package_dir STREQUAL "")
  _lcm_add_outputs(_JAVA_SOURCES lcmtypes/${_type}.java)
else()
  _lcm_add_outputs(_JAVA_SOURCES ${_package_dir}/${_type}.java)
endif()
```
Oh, and... I just realized, I think `_package_dir` should be reset to `"."`, not `""`? Otherwise it looks like you end up with absolute paths where you don't want them. (This *definitely* makes me think the 'no package specified' path is poorly tested :slightly_smiling_face:.) Note that this means changing the above `if`, also.

Reply:
I fully agree with your comment!
Funny thing is, that is what I had before I tried to make it less verbose/repetitive 🙂. Instead I ended up with a prettier but more naïve implementation that could introduce inter-language leaks instead.

I see that the changes have been approved, but I think it's best I update it.

Also, resetting `_package_dir` to `"."` instead isn't a bad idea. I suppose it won't make any functional difference since `_DESTINATION` is set to `${CMAKE_CURRENT_BINARY_DIR}` if it's not defined, and if I'm not mistaken, all paths are prepended with `_DESTINATION` (or `_GO_DESTINATION`) before they are used.
Nevertheless, resetting to `"."` seems more robust and will also make sure there aren't any paths containing consecutive slashes stored in `_CPP_HEADERS`.